// @google-cloud/monitoring/build/src/v3/index.js
"use strict";
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by gapic-generator-typescript. **
// ** https://github.com/googleapis/gapic-generator-typescript **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.UptimeCheckServiceClient = exports.SnoozeServiceClient = exports.ServiceMonitoringServiceClient = exports.QueryServiceClient = exports.NotificationChannelServiceClient = exports.MetricServiceClient = exports.GroupServiceClient = exports.AlertPolicyServiceClient = void 0;
var alert_policy_service_client_1 = require("./alert_policy_service_client");
Object.defineProperty(exports, "AlertPolicyServiceClient", { enumerable: true, get: function () { return alert_policy_service_client_1.AlertPolicyServiceClient; } });
var group_service_client_1 = require("./group_service_client");
Object.defineProperty(exports, "GroupServiceClient", { enumerable: true, get: function () { return group_service_client_1.GroupServiceClient; } });
var metric_service_client_1 = require("./metric_service_client");
Object.defineProperty(exports, "MetricServiceClient", { enumerable: true, get: function () { return metric_service_client_1.MetricServiceClient; } });
var notification_channel_service_client_1 = require("./notification_channel_service_client");
Object.defineProperty(exports, "NotificationChannelServiceClient", { enumerable: true, get: function () { return notification_channel_service_client_1.NotificationChannelServiceClient; } });
var query_service_client_1 = require("./query_service_client");
Object.defineProperty(exports, "QueryServiceClient", { enumerable: true, get: function () { return query_service_client_1.QueryServiceClient; } });
var service_monitoring_service_client_1 = require("./service_monitoring_service_client");
Object.defineProperty(exports, "ServiceMonitoringServiceClient", { enumerable: true, get: function () { return service_monitoring_service_client_1.ServiceMonitoringServiceClient; } });
var snooze_service_client_1 = require("./snooze_service_client");
Object.defineProperty(exports, "SnoozeServiceClient", { enumerable: true, get: function () { return snooze_service_client_1.SnoozeServiceClient; } });
var uptime_check_service_client_1 = require("./uptime_check_service_client");
Object.defineProperty(exports, "UptimeCheckServiceClient", { enumerable: true, get: function () { return uptime_check_service_client_1.UptimeCheckServiceClient; } });
